src/execution/execute.ts

```diff
@@ -210,12 +268,58 @@ export interface ExecutionArgs {
  *
  * If the arguments to this function do not result in a legal execution context,
  * a GraphQLError will be thrown immediately explaining the invalid input.
+ *
+ * This function does not support incremental delivery (`@defer` and `@stream`).
+ * If an operation which would defer or stream data is executed with this function,
+ * it will throw or reject instead. Use `experimentalExecuteIncrementally` if you
+ * want to support incremental delivery.
  */
-export function execute(
+export function execute(args: ExecutionArgs): PromiseOrValue<ExecutionResult> {
+  const result = executeSyncOrIncrementally(args);
+  if (!isPromise(result)) {
+    return result;
+  }
+
+  return result.then((incrementalResult) => {
+    if ('singleResult' in incrementalResult) {
+      return incrementalResult.singleResult;
+    }
+    throw new GraphQLError(
+      'Executing this GraphQL operation would unexpectedly produce multiple payloads (due to @defer or @stream directive)',
+    );
+  });
+}
+
+/**
+ * Implements the "Executing requests" section of the GraphQL specification,
+ * including `@defer` and `@stream` as proposed in
+ * https://github.com/graphql/graphql-spec/pull/742
+ *
+ * This function returns a Promise of an ExperimentalExecuteIncrementallyResults
+ * object. This object either contains a single ExecutionResult as
+ * `singleResult`, or an `initialResult` and a stream of `subsequentResults`.
+ *
+ * If the arguments to this function do not result in a legal execution context,
+ * a GraphQLError will be thrown immediately explaining the invalid input.
+ */
+export function experimentalExecuteIncrementally(
   args: ExecutionArgs,
-): PromiseOrValue<
-  ExecutionResult | AsyncGenerator<AsyncExecutionResult, void, void>
-> {
+): Promise<ExperimentalExecuteIncrementallyResults> {
+  const result = executeSyncOrIncrementally(args);
+  if (isPromise(result)) {
+    return result;
+  }
+  // Always return a Promise for a consistent API.
+  return Promise.resolve({ singleResult: result });
```

Comment:
Question: Is the benefit of always returning a promise definitely worth it?

1. If the first result is synchronous, 
returning the generator could also be synchronous and then we save a tick.

2. This change actually sort of makes the API inconsistent in another sense as non incremental delivery can be either value or Promise, so maybe shouldn't incremental.

3. I believe we recently changed in v17 subscribe to allow returning a generator synchronously. Looks like now we r headed in the other direction. Either way, should they be aligned?

Reply:
@yaacovCR  Agree, with can make this function return `PromiseOrValue<ExperimentalExecuteIncrementallyResults>`